### PR TITLE
(Gh 7971) Can define discriminators in the schema

### PR DESCRIPTION
**Pull request by IslandRhythms**

### CODE REVIEW

File: lib/schema.js [hunk 1]
```diff
@@ -569,6 +575,12 @@ Schema.prototype.add = function add(obj, prefix) {
           this.nested[prefix.substring(0, prefix.length - 1)] = true;
         }
         this.path(prefix + key, val);
+        if (val.discriminators) {
```
Comment (vkarpov15) [inline lib/schema.js:578]: There's a case we missed: `val` can be a Schema, and Schemas can have a `discriminators` property. Let's replace this with `if (utils.isPOJO(val) && utils.isPOJO(val.discriminators))`

File: lib/schema.js [hunk 2]
```diff
@@ -540,6 +540,12 @@ Schema.prototype.add = function add(obj, prefix) {
         this.nested[prefix.substring(0, prefix.length - 1)] = true;
       }
       this.path(prefix + key, val);
+      if (Array.isArray(val) && val.length != 0 && val[0].discriminators) {
```
Comment (vkarpov15) [inline lib/schema.js:543]: There's a case we missed: `val` can be a Schema, and Schemas can have a `discriminators` property. Let's replace this with `if (utils.isPOJO(val[0]) && utils.isPOJO(val[0].discriminators))`

**Review by vkarpov15 — CHANGES_REQUESTED** [review]

File: lib/schema.js [hunk 3]
```diff
@@ -540,6 +540,12 @@ Schema.prototype.add = function add(obj, prefix) {
         this.nested[prefix.substring(0, prefix.length - 1)] = true;
       }
       this.path(prefix + key, val);
+      if (utils.isPOJO(val[0]) && utils.isPOJO(val[0].discriminators)) {
```
Comment (AbdelrahmanHafez) [inline lib/schema.js:543]:
```suggestion
      if (utils.isPOJO(val[0] && val[0].discriminators)) {
```

Comment (AbdelrahmanHafez) [reply lib/schema.js:543]: @vkarpov15 
Would it be more intention revealing to use
`if ((val[0] instanceof Schema) === false && val[0].discriminators) {`?

Comment (vkarpov15) [reply lib/schema.js:543]: That's fair, we can do `&& !(val[0] instanceof Schema)` instead

**Pull request merged**